### PR TITLE
[2.0a2] [retiarii] improvement

### PR DESCRIPTION
**Pull request by QuanluZhang**
- [x] support different pytorch versions (pytorch >=1.5.0)
- [x] refactor of execution engine, execution engine is instantiated in RetiariiAdvisor so that the engine can receive all the events
- [x] tpe strategy supports concurrency
- [x] unify user interface for classic non-weight-sharing search and weight-sharing search
- [x] support darts weight sharing trainer with the unified interface
- [x] merge new changes in new experiment launching to RetiariiExperiment
- [x] support `n_chosen > 1` in InputChoice
- [x] support random exploration strategy
- [ ] support enas example, and decorator `@stop_parsing_module()`

### CODE REVIEW

File: nni/retiarii/nn/pytorch/nn.py [hunk 1]
```diff
@@ -29,12 +32,18 @@
...
     'AdaptiveLogSoftmaxWithLoss', 'TransformerEncoder', 'TransformerDecoder',
     'TransformerEncoderLayer', 'TransformerDecoderLayer', 'Transformer',
-    #'LazyLinear', 'LazyConv1d', 'LazyConv2d', 'LazyConv3d',
-    #'LazyConvTranspose1d', 'LazyConvTranspose2d', 'LazyConvTranspose3d',
-    #'Unflatten', 'SiLU', 'TripletMarginWithDistanceLoss', 'ChannelShuffle',
-    'Flatten', 'Hardsigmoid', 'Hardswish'
+    'Flatten', 'Hardsigmoid'
 ]
 
+if version.parse(torch.__version__) >= version.parse('1.6.0'):
```
Comment (ultmaster) [inline nni/retiarii/nn/pytorch/nn.py:38]: Is this "the" correct checking? I saw many occurrences in our repo where:

image

File: nni/retiarii/execution/base.py [hunk 2]
```diff
@@ -93,18 +95,21 @@ def _final_metric_callback(self, trial_id: int, metrics: MetricData) -> None:
...
+        return self.resources
 
     @classmethod
     def trial_execute_graph(cls) -> None:
         """
         Initialize the model, hand it over to trainer.
         """
         graph_data = BaseGraphData.load(receive_trial_parameters())
-        with open('_generated_model.py', 'w') as f:
+        random_str = ''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(6))
```
Comment (hzhua) [inline nni/retiarii/execution/base.py:107]: We may check if the random_str is repeated, although the probability is low.
A lock is needed if trial_execute_graph can be executed concurrently.

Comment (QuanluZhang) [reply nni/retiarii/execution/base.py:107]: yes, we can check that.
why a lock is required?

File: nni/retiarii/execution/base.py [hunk 3]
```diff
@@ -93,18 +95,21 @@ def _final_metric_callback(self, trial_id: int, metrics: MetricData) -> None:
...
+        file_name = f'_generated_model_{random_str}.py'
```
Comment (ultmaster) [inline nni/retiarii/execution/base.py:108]: Might be better if we put all temp files into a folder called `generated` or `.retiarii` or something...

Comment (QuanluZhang) [reply nni/retiarii/execution/base.py:108]: good point, will update

File: nni/retiarii/execution/base.py (hunk 2 above)
Comment (hzhua) [reply nni/retiarii/execution/base.py:107]: We may need a set to check the existed model names. If two models are generated concurrently in two threads and they generate the same random_str (although with a very very low probability), the random_str is not in the set when checking its existence. Then both models will have the same random_str.

Comment (ultmaster) [reply nni/retiarii/execution/base.py:107]: Another way out is to decrease the probability. With a sufficiently long hexdigit (e.g., 16 characters), this is almost impossible.

**Comment by liuzhe-lz** [comment]
Is this targetting v2.0 release?

**Comment by QuanluZhang** [comment]
> Is this targetting v2.0 release?

yes, better to be included in v2.0

File: nni/retiarii/nn/pytorch/nn.py [hunk 4]
```diff
@@ -5,10 +5,12 @@
 import torch
 import torch.nn as nn
 
-from ...utils import add_record
+from ...utils import add_record, version_larger_equal
```
Comment (ultmaster) [inline nni/retiarii/nn/pytorch/nn.py:8]: nn.pytorch.nn looks strange. Why not nn.pytorch?

**Comment by liuzhe-lz** [comment]
Merge before alpha2

File: nni/retiarii/experiment.py [hunk 5]
```diff
@@ -156,42 +169,79 @@ def start(self, config: RetiariiExeConfig, port: int = 8080, debug: bool = False
...
+            self._dispatcher_thread.join(timeout=1)
 
         self.port = None
         self._proc = None
         self._pipe = None
+        self._dispatcher = None
+        self._dispatcher_thread = None
 
-    def run(self, config: RetiariiExeConfig, port: int = 8080, debug: bool = False) -> str:
+    def run(self, config: RetiariiExeConfig = None, port: int = 8080, debug: bool = False) -> str:
```
Comment (liuzhe-lz) [inline nni/retiarii/experiment.py:209]: This looks identical to base class. Why overriding?

File: nni/retiarii/execution/base.py (hunk 2 above)
Comment (QuanluZhang) [reply nni/retiarii/execution/base.py:107]: good suggestions, i will fix this one in the next pr

File: nni/retiarii/execution/base.py (hunk 3 above)
Comment (QuanluZhang) [reply nni/retiarii/execution/base.py:108]: in the next pr

File: nni/retiarii/experiment.py (hunk 5 above)
Comment (QuanluZhang) [reply nni/retiarii/experiment.py:210]: this one is not identical, because this is `OneShotTrainers`

Comment (QuanluZhang) [reply nni/retiarii/experiment.py:210]: `get_status` can be removed

Comment (liuzhe-lz) [reply nni/retiarii/experiment.py:210]: Copied part should be `super().run(...)`

Comment (QuanluZhang) [reply nni/retiarii/experiment.py:210]: update, thanks

Comment (liuzhe-lz) [reply nni/retiarii/experiment.py:210]: It also applies to `start` and `stop`...

Comment (QuanluZhang) [reply nni/retiarii/experiment.py:210]: yeah, `stop` can be removed. but `start` cannot, because dispatcher is instantiated in it.

**Review by ultmaster — APPROVED** [review]

**Review by ultmaster — APPROVED** [review]

**Review by ultmaster — APPROVED** [review]

**Review by liuzhe-lz — APPROVED** [review]

**Pull request merged**